### ISSUE TITLE
lunar_lander: widen starts, train/validate split, time-boxed evolve

### ISSUE THREAD

**Issue by nleck** [stSoftwareAU/NEAT-AI-Examples#193]
There appears to be no need to do anything other than hit the trust at a particular point 

True model should be able to handle different situations so much so I think we've cheated.

Can we show the same final creature handling several different random starting scenarios? To prove learning not remembering 

It's OK to evolve for a while to get a creature capable of landing successfully the bulk of the time. 

We don't generate the screen shots on each build or anything like that.

Let's take an option of time to evolve for in minutes for our example script ( quality/CI should be every fast ). I'm happy for the examples to take minutes to run.

We should record/chart the evolution process, we should follow the stand approaches of training with many scenarios and test/validate with others . 

The SVGs should be of the validation run.

Again seperate the CI/quality ( very fast) from the runnable training/test/validate 

I can see this with

https://github.com/stSoftwareAU/NEAT-AI-Examples/blob/Develop/lunar_lander/README.md

But I suspect same for other examples, happy for the examples training/test/validate to take minutes ( controlled by user). If they want to train for a minute fine you'll get what we can evolve in one minute, they want to evolve for five minutes then they should get better results. We can pass target error with a timeout of X minutes ( standard options of NEAT-AI) no complex logic in the example script.

Target error much better, no point wasting minutes if we've hit the target

Maybe we default to 1% error which user can change and 2 minutes evolution time by default ( excluding the pre/post work after evolving) 

We need to update the articles with the actual statistic, N generations in Y time resulting an error of Z%

## Current Understanding

The `lunar_lander` example currently cheats: starts are narrow enough that one canned thrust pulse wins, so the champion is remembering one trajectory rather than learning to land. We will widen the start distribution, split scenarios into a training set and a held-out validation set, render the SVGs from the validation run, and drive evolution with NEAT-AI's native `targetError` + `timeoutMinutes` options (no bespoke loop logic) — whichever limit hits first wins, and in practice target error is expected to hit first. As part of this PR the worker runs the example once, captures the real run statistics, and writes them into the `lunar_lander/README.md` (and any related article) so the documentation reflects what actually happened. No auto-injection mechanism is built — future developer runs do not modify the docs. Other examples get follow-up issues once the pattern is proven.

### Accepted scope so far
- First PR is `lunar_lander` only; follow-up issues track the same pattern for other examples.
- Target metric: **landed-rate >= 99% on the validation set** (i.e. error = 1 - landed_rate, so the existing 1% default still reads naturally).
- Many randomised scenarios for both training and validation: different start positions, velocities, tilt, and randomised pad position. Always recoverable (no impossible starts).
- Validation scenarios are drawn from the same wider distribution but use disjoint random seeds, so a champion that memorises training cannot pass validation.
- Evolution uses NEAT-AI's standard `targetError` + `timeoutMinutes` options. Defaults: 1% target error, 2 minutes wall-clock, both user-overridable. Either limit ends evolution; target-error is the expected common case.
- **Scenario counts: 1000 training, 200 validation.** User accepted that this may need more wall-clock than the 2-minute default; users who want more variety can raise `timeoutMinutes`.
- CI/quality stays very fast via a tiny budget run (5-10 seconds) that just proves the pipeline wires up; runnable training/test/validate is a separate path that may take minutes.
- README/article updates are a one-shot, manual-by-the-worker step in this PR. The worker performs a real training run, reads the captured statistics, and writes the actual numbers (N generations in Y seconds reaching Z% landed-rate) into the docs. There is no `<!-- STATS -->` block, no `last-run.json` artifact, and no per-developer-run rewrite of the docs.
- **Evolution telemetry must give good insights into how fitness improves over generations and demonstrate the final result.** Concretely: per-generation CSV of best/avg fitness, plus a rendered SVG line chart of best/avg fitness vs generation embedded in the README, plus a per-validation-scenario landed-rate bar chart. The user said "happy for more" twice and emphasised illustrating the evolution process and demonstrating the result — the comprehensive option matches that intent.
- The chart/recording pattern is intended to apply to all examples eventually; `lunar_lander` is the proving ground.

### Open questions
None.

### Assumptions
- "Good insights into how fitness improves over generations" is satisfied by per-generation best/avg fitness in CSV + an embedded SVG line chart. The per-validation-scenario landed-rate bar chart covers "demonstrate the result". If the user wants additional series (population diversity, species count, mutation rates), they will say so on a later issue.
- The fast CI path runs the existing script with a 5-10 second budget; it does not load a checked-in champion. (User picked the tiny-budget option, not the champion-JSON option.)
- "Randomised pad position" means the pad x-coordinate varies per scenario but the pad stays on the ground plane and remains reachable from any starting state.
- Validation seeds are fixed (deterministic) so a re-run produces the same SVG, but they are distinct from the training seed pool.
- Because the README update is a one-time worker action in this PR, no convention (`<!-- STATS -->` markers, `last-run.json`) is introduced for sibling docs issues to adopt.

### Related open issues
- #191 — Articles overstate generations/time vs what actually happened (`mnist_classification` specifically). Same root cause as this issue: READMEs not driven by real run stats. Resolving #193 fixes the `lunar_lander` instance directly; #191 will still need its own pass for `mnist_classification`.
- #190, #189, #188, #187, #186 — all README/documentation passes on examples. No direct conflict. Since #193 will not introduce a stats-injection convention, these issues remain free to set their own README structure.

**Comment by stservice** [comment]
## Grill-Me Round 1

**TL;DR** — A few choices needed before we widen the starts, add a train/validate split, and switch to time-boxed evolution.

### Understanding
Examples currently cheat: starts are narrow enough that one canned thrust wins. Plan: widen the start distribution, split scenarios into train + held-out validate, render SVGs from validation, and drive evolution with NEAT-AI's native target-error + timeout-in-minutes options (default 1% / 2 min, user-overridable). READMEs get updated with real run stats (related: #191).

### Questions

1. Scope for the first PR?
   - [x] Just `lunar_lander` (prove the pattern, then issue follow-ups for each other example)
   - [ ] All examples in one sweep
   - [ ] `lunar_lander` + `mnist_classification` together (pairs with #191)
   - [ ] other — please describe in a reply

2. What does "1% error" mean for `lunar_lander` (no classification error)?
   - [ ] Treat `1 - landed_rate` as error → target landed-rate ≥ 99% on validation
   - [ ] Use a softer landing-rate target (e.g. ≥ 90%) and document per-example targets
   - [ ] Keep 1% as the generic default; per-example metric defined in each script
   - [x] other — please describe in a reply

3. How much harder should the starting distribution be?
   - [x] Wider spawn box + larger initial velocity + tilt jitter (still always recoverable)
   - [x] Same as above, plus randomised pad position
   - [ ] Aggressive: occasional unrecoverable starts allowed (champion just has to do well on average)
   - [ ] other — please describe in a reply

4. How does CI/quality stay fast?
   - [ ] Skip evolution entirely; load a checked-in champion JSON and assert it still passes validation
   - [x] Tiny budget run (e.g. 5–10 seconds) just to prove the pipeline wires up
   - [ ] Both — checked-in champion for behaviour, tiny budget for plumbing
   - [ ] other — please describe in a reply

5. README stats — auto-injected or manual?
   - [ ] Runner writes a `<!-- STATS -->` block into the README on each training run
   - [ ] Runner writes `docs/<example>/last-run.json`; README links/embeds it
   - [ ] Manual update by the developer after a training run
   - [x] other — please describe in a reply

Tick the boxes that apply (the GitHub mobile app lets you tap each one), then add any free-form notes in a reply.

**⏳ Awaiting your reply.** The Vibe Coder is waiting for your response on this issue to continue grilling.

**Comment by nleck** [comment]
2. I would have thought that an 1% error would mean that we landed OK 99% of the time. Remember I want lots of senarios ( different start positions, speed etc ) 

I want the examples to LEARN not  remember... I suspect I'm seeing remembering.

**Comment by stservice** [comment]
## Grill-Me Round 2

**TL;DR** - Three loose ends before we converge: the README stats mechanism, scenario counts, and what "record/chart the evolution" should produce.

### Understanding
Settled so far: `lunar_lander` first; target = landed-rate >= 99% on a held-out validation set; wider starts plus randomised pad; tiny CI budget (5-10 s); `targetError` + `timeoutMinutes` drive evolution (defaults 1% / 2 min). Validation seeds are disjoint from training so the champion has to learn, not remember.

### Questions

1. README stats injection (you picked "other" last round - which way do you want it?)
   - [ ] Runner writes a `<!-- STATS -->` block straight into `lunar_lander/README.md` on each training run
   - [ ] Runner writes `lunar_lander/last-run.json`; README links/embeds it
   - [ ] Runner writes both a `<!-- STATS -->` block and the JSON
   - [ ] Manual update by the developer after a training run
   - [x] other - please describe in a reply

2. How many scenarios for training and validation?
   - [ ] Train 100, validate 25 (quick to run, modest signal)
   - [ ] Train 500, validate 100 (good signal, fits in 2 minutes)
   - [x] Train 1000, validate 200 (lots of variety, may need more wall-clock)
   - [ ] Configurable via CLI flag, default 500/100
   - [ ] other - please describe in a reply

3. What does "record/chart the evolution process" produce?
   - [ ] Per-generation CSV of best/avg fitness only
   - [ ] CSV plus a rendered SVG line chart embedded in the README
   - [x] CSV plus SVG plus a final landed-rate-per-validation-scenario bar chart
   - [ ] other - please describe in a reply

Tick the boxes that apply (the GitHub mobile app lets you tap each one), then add any free-form notes in a reply.

**⏳ Awaiting your reply.** The Vibe Coder is waiting for your response on this issue to continue grilling.

**Comment by nleck** [comment]
5. You are record/charting the statistics now. The documentation we are doing now wouldn't be affected by the developer running the examples. To do the documentation you would run the example and take the statistics calculated and update the documents/articals.

**Comment by stservice** [comment]
## Grill-Me Round 3

**TL;DR** — One thing left: what should the "record/chart the evolution" output look like?

### Understanding
Settled: `lunar_lander` only; target landed-rate >= 99% on a held-out validation set; wider starts plus randomised pad; 1000 training / 200 validation scenarios with disjoint seeds; tiny CI budget (5-10s); `targetError` + `timeoutMinutes` drive evolution (defaults 1% / 2 min, user-overridable); the worker runs the example once during this PR and writes the real numbers into the README — no auto-injection plumbing.

### Questions

1. What should the evolution recording produce?
   - [x] Per-generation CSV of best/avg fitness only (simplest)
   - [ ] CSV plus a rendered SVG line chart of best/avg fitness vs generation, embedded in the README
   - [ ] CSV plus the line-chart SVG plus a per-validation-scenario landed-rate bar chart
   - [ ] other — please describe in a reply

Tick the box that applies (the GitHub mobile app lets you tap it), then add any free-form notes in a reply.

**⏳ Awaiting your reply.** The Vibe Coder is waiting for your response on this issue to continue grilling.

**Comment by nleck** [comment]
Round 2 Q1: You're recording/charts the stats now 

Round 2 Q2: I think you'll find that we can scan over a lot of training data quickly. I want all this recorded as part of the statistics  reported. 

Round 2 Q3: You're recording/charting the evolution stats now, I'm happy for more. I want to illistrate the evolution process and then demostrate the result. This would be for all examples.

**Comment by nleck** [comment]
The evolution statistics should record what's needed to give good insights to how the fitness improves over the generations. 

We will have two limits. The target error rate and the wall clock, many times we will hit the target error rate first

**Comment by stservice** [comment]
## Grill-Me — Ready for Next Phase

**TL;DR** — `lunar_lander` rewrite: wider randomised starts, 1000/200 train/validate split with disjoint seeds, native `targetError` + `timeoutMinutes` (default 1% / 2 min, either limit wins), tiny CI budget, comprehensive evolution telemetry, README rewritten with real run stats during this PR.

### Agreed problem statement
The current `lunar_lander` champion is remembering one canned trajectory because the start distribution is too narrow. Fix: widen starts, force the champion to generalise via a held-out validation set, drive evolution with NEAT-AI's native limits, and have the worker bake the real run statistics into the docs as a one-off in this PR.

### Accepted scope
- First PR is `lunar_lander` only. Other examples get follow-up issues once the pattern is proven.
- Target metric: landed-rate ≥ 99% on validation (i.e. `error = 1 - landed_rate`, defaulting to 1%).
- 1000 training scenarios, 200 validation scenarios; disjoint random seeds; randomised start position, velocity, tilt, and pad position; all recoverable.
- Evolution uses NEAT-AI's standard `targetError` + `timeoutMinutes`. Defaults 1% / 2 minutes, both user-overridable. Either limit ends the run; target-error is the expected common case.
- CI/quality path: tiny 5–10 second budget run to prove the pipeline wires up. Separate from the runnable training/validate path.
- Evolution telemetry: per-generation CSV of best/avg fitness + embedded SVG line chart of best/avg vs generation + per-validation-scenario landed-rate bar chart.
- README/article: worker runs the example once during this PR and writes the real numbers (N generations in Y seconds at Z% landed-rate) into `lunar_lander/README.md`. No `<!-- STATS -->` block, no `last-run.json`, no per-developer-run rewrite.

### Explicit assumptions
- The CSV + line chart + bar chart combination is enough to "give good insights" and "demonstrate the result". Extra series (diversity, species count) are out of scope unless raised separately.
- Validation seeds are fixed/deterministic so the SVG is reproducible.
- Randomised pad x-coordinate stays on the ground plane and reachable from every starting state.

### Related open issues
- #191 — `mnist_classification` README has the same overstated-stats problem. Out of scope here; will need its own pass.
- #190, #189, #188, #187, #186 — README documentation passes on other examples. No direct conflict; #193 deliberately does not introduce a stats-injection convention, so they stay free to choose their own structure.

### How would you like to proceed?
- [ ] Apply the planning label
- [ ] Apply the work-on label

Tick the box for your choice, then open the issue on GitHub mobile, tap the labels icon, and toggle either `planning` or `work-on` to actually apply it.

**Comment by stservice** [comment]
Claimed by `stservice-1778290948692`

**Comment by stservice** [comment]
## Plan: 7 sub-issues, ordered by dependency

The work breaks naturally into seven tasks, each sized for a single PR. Foundation lands first, the README refresh closes the loop once everything else is in place.

### Sub-issues created

| # | Title | Depends on |
| --- | --- | --- |
| #195 | Widen scenario distribution + disjoint train/validate seed pools | — |
| #196 | Replace generation cap with `targetError` + `timeoutMinutes` stop conditions | #195 |
| #198 | Validation pipeline + validation-driven SVG | #195, #196 |
| #199 | Per-generation evolution CSV + best/avg fitness line chart | #196 |
| #200 | Per-validation-scenario landed-rate bar chart | #198 |
| #201 | Fast CI/quality budget separate from runnable training/validate | #196 |
| #202 | Refresh README with real run statistics + new pipeline | #195, #196, #198, #199, #200, #201 |

### Dependency graph

```mermaid
flowchart LR
    A[#195 widen + train/val split] --> B[#196 time-boxed evolve]
    A --> C[#198 validation + SVG]
    B --> C
    B --> D[#199 CSV + fitness chart]
    C --> E[#200 validation bar chart]
    B --> F[#201 fast CI path]
    A --> G[#202 README refresh]
    B --> G
    C --> G
    D --> G
    E --> G
    F --> G
```

### Suggested implementation order

1. **#195** — foundation, no deps. Wider sampler + disjoint seed pools.
2. **#196** — swap stop conditions to `targetError`/`timeoutMinutes`. After this lands, #199 and #201 can run in parallel.
3. **#198, #199, #201** — three parallel branches once #196 is in.
4. **#200** — needs validation results from #198.
5. **#202** — closes the loop after every other sub-issue is merged.

### Assumptions and decisions made during planning

- **Stop conditions live in the existing hand-rolled loop**, not a switch to `creature.evolveDir`. The lunar-lander task is RL-style, so a custom fitness function is unavoidable. The user's "standard options of NEAT-AI" maps to using the same field names (`targetError`, `timeoutMinutes`) on `EvolveOptions` rather than rewriting onto `evolveDir` — that interpretation matches "no complex logic in the example script".
- **`targetError = 0.01`** maps to **landed-rate ≥ 99% on the training trial batch**. Encoding `error = 1 - landed_rate` reads naturally and matches the user's accepted scope.
- **Validation SVG selection rule**: median-fitness validation scenario by default; if all 200 land, scenario at index 0. Documented in JSDoc on `validateChampion`.
- **Chart utilities live under `common/`** so other examples can adopt the pattern after lunar-lander proves it (the user said this is the proving ground).
- **Fast CI path**: tiny budget (≈6s timeout, impossible target error) — the user-preferred option, not a checked-in champion JSON.
- **No `<!-- STATS -->` markers, no `last-run.json`, no auto-injection.** README refresh is a one-shot manual edit per the user's explicit guidance.
- **#191 (mnist_classification overstates generations)** has the same root cause but a different example. #193 does **not** introduce a convention #191 would inherit, so they remain independent passes.

### A note on this planning issue

The `planning` label will be removed and this issue closed automatically once planning is recorded. If any of the breakdown above needs adjusting — e.g., a different validation SVG selection rule, a different chart layout, merging two sub-issues — **reopen this issue** with the change you want and I'll iterate on the plan.

**Comment by stservice** [comment]
## Planning Complete

Planning complete. **7 sub-issue(s)** created:

- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/202
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/201
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/200
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/199
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/198
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/196
- https://github.com/stSoftwareAU/NEAT-AI-Examples/issues/195

---
🤖 Processed by: stservice

**Comment by stservice** [comment]
Planning complete — sub-issues created.
